### ISSUE TITLE
RFC: shared data holder for loops

### ISSUE THREAD

**Issue by awaelchli** [PyTorchLightning/pytorch-lightning#9220]
## Data holder for loops

Provide a means to share data in a stack of loops, for example in the training loop which is made of
FitLoop > TrainingEpochLoop > TrainingBatchLoop > OptimizerLoop, where "X > Y"  denotes Y is a child loop of X. 

### Motivation

We currently have several attributes like running_loss, skip_backward, etc. which need to be accessed externally or from within subloops.
The current way is to reach into these loops like so:

`self.trainer.fit_loop.epoch_loop.batch_loop.running_loss`

If a subloop needs to access a parent loop's attribute like this, then it has several drawbacks.
- Restricted to the current loop structure, less flexible for loop customization
- A child loop requires knowlege of the implementation in the parent. This makes it harder to provide backward compatibility. 

### Pitch

Introduce a dataclass or dictionary:

```python
@dataclass
class LoopData:
    """The attributes here are shared across all loops"""
    global_step: int = 0
    running_loss: Tensor = tensor(0.)
    accumulated_loss: Tensor = tensor(0.)
    skip_backward: bool = False

class ParentLoop(Loop):
    child_loop = ChildLoop()

    def foo(self):
        print(self.shared_data.running_loss)
        
class ChildLoop(Loop):
    def bar(self):
        print(self.shared_data.running_loss)  # shared across parent and this child
```

(the attributes here are for illustrative purpose, whether they get included or not is for discussion)

This approach may still be too strict and it could be extended to make the data holder dynamic and let each loop register attributes at the time of instantiation. 

### Additional context

We are currently extracting an optimizer loop in #9191 which is the origin of this RFC. There we are splitting the code path of manual and automatic optimization. In these cases, the updating logic for the attributes [running loss and accumulated_loss](https://github.com/PyTorchLightning/pytorch-lightning/blob/9701c8949bebb98afd79a61ba2b866d1f8359990/pytorch_lightning/loops/optimizer/optimizer_loop.py#L159)
should also be different but currently these attributes are owned by the `TrainingBatchLoop`. A data holder would greatly improve the situation there.
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

**Comment by awaelchli** [comment]
@tchaton @ananthsub @carmocca

**Comment by tchaton** [comment]
Hey @awaelchli,

Quick questions to better understand the background:

* Do you see anything else the loops might want to communicate between each other ?
* Should we limit accessibility from parent to its children ? 
* Would it make sense to have a flatten loop state only to external components such a logging ?
* Do you see anything in the state which could require to be saved within the checkpoint ?

Best,
T.C

**Comment by carmocca** [comment]
As it stands, all the instances where we need to access shared data are points where the code could be improved, perhaps with the exception of `global_step` which is really a global value to track.

This is why I don't think this refactor is important right now.

**Comment by ananthsub** [comment]
> As it stands, all the instances where we need to access shared data are points where the code could be improved, perhaps with the exception of `global_step` which is really a global value to track.

@carmocca i'm curious to hear your ideas on this! needing access to the trainer to walk down the loop hierarchy to access state somewhere is very roundabout right now.

i also wonder: how easy is it to read this code now vs having dedicated implementations of training loops that can be subbed into the fit loop to run. it seems like this problem arises only in the training loops compared to the evaluation and prediction loops, given the differences in training modes lightning supports. 

if we have pre-packaged loop implementations for training, the code/state is all in one place. the somewhat hacky part is needing a loop selector to pick the training loop implementation based on the lightning module properties. 

vs.with fit_loop.epoch_loop.batch_loop.optimizer_loop this is hopping across 4 different classes to trace through the execution.
